meta/client.go

```diff
@@ -0,0 +1,216 @@
+		ZkConn: zkConn,
+		Tables: tables,
+		Metas:  make(map[string]*session.MetaManager),
+	}
+}
+
+func (m *ClusterManager) getMeta(table string) (*session.MetaManager, error) {
+	var meta *session.MetaManager
+
+	tableInfo, err := globalClusterManager.Tables.Get(table)
+	if err == nil {
+		meta = globalClusterManager.Metas[tableInfo.(*TableInfoWatcher).metaAddrs]
+		if meta != nil {
+			return meta, nil
+		}
+	}
+
+	logrus.Printf("[%s] can't get cluster info from local cache, try fetch from zk.", table)
+	globalClusterManager.Mut.Lock()
+	defer globalClusterManager.Mut.Unlock()
+	tableInfo, err = globalClusterManager.Tables.Get(table)
+	if err != nil {
+		tableInfo, err = m.getTableInfo(table)
+		if err != nil {
+			logrus.Errorf("get table[%s] info failed: %s", table, err)
+			return nil, err
+		}
+		err = globalClusterManager.Tables.Set(table, tableInfo)
+		if err != nil {
+			logrus.Warnf("[%s] cluster info update local cache failed: %s", table, err)
+		}
+	}
+
+	metaAddrs := tableInfo.(*TableInfoWatcher).metaAddrs
+	meta = globalClusterManager.Metas[metaAddrs]
+	if meta == nil {
+		metaList, err := parseToMetaList(metaAddrs)
+		if err != nil {
+			logrus.Errorf("The table[%s] cluster addr[%s] format is err: %s", table, metaAddrs, err)
+			return nil, base.ERR_INVALID_DATA
+		}
+		meta = session.NewMetaManager(metaList, session.NewNodeSession)
+		globalClusterManager.Metas[metaAddrs] = meta
+	}
+
+	return meta, nil
+}
+
+// get table cluster info and watch it based table name from zk
+// The zookeeper path layout:
+// /<RegionPathRoot>/<table> =>
+//                         {
+//                           "cluster_name" : "clusterName",
+//                           "meta_addrs" : "metaAddr1,metaAddr2,metaAddr3"
+//                         }
+func (m *ClusterManager) getTableInfo(table string) (*TableInfoWatcher, error) {
+	path := fmt.Sprintf("%s/%s", zkRoot, table)
+	value, _, watcherEvent, err := globalClusterManager.ZkConn.GetW(path)
+	if err != nil {
+		if err == zk.ErrNoNode {
+			logrus.Errorf("the table[%s] info doesn't exist on zk[%s], err: %s", table, path, err)
+			return nil, base.ERR_OBJECT_NOT_FOUND
+		} else {
+			logrus.Errorf("get table[%s] info from zk[%s] failed: %s", table, path, err)
+			return nil, base.ERR_ZOOKEEPER_OPERATION
+		}
+	}
+
+	type clusterInfoStruct struct {
+		Name      string `json:"cluster_name"`
+		MetaAddrs string `json:"meta_addrs"`
+	}
+	var cluster = &clusterInfoStruct{}
+	err = json.Unmarshal(value, cluster)
+	if err != nil {
+		logrus.Errorf("table[%s] info on zk[%s] format is invalid, err = %s", table, path, err)
+		return nil, base.ERR_INVALID_DATA
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	tableInfo := TableInfoWatcher{
+		tableName:   table,
+		clusterName: cluster.Name,
+		metaAddrs:   cluster.MetaAddrs,
+		event:       watcherEvent,
+		ctx: zkContext{
+			ctx:    ctx,
+			cancel: cancel,
+		},
+	}
+	go watchTableInfoChanged(tableInfo)
+
+	return &tableInfo, nil
```

Comment:
```suggestion
	tableInfo := &TableInfoWatcher{
		tableName:   table,
		clusterName: cluster.Name,
		metaAddrs:   cluster.MetaAddrs,
		event:       watcherEvent,
		ctx: zkContext{
			ctx:    ctx,
			cancel: cancel,
		},
	}
	go watchTableInfoChanged(tableInfo)

	return tableInfo, nil
```

you should pass the pointer of `tableInfo` to `watchTableInfoChanged`. Otherwise the returned one, and the "watching" one, will not be the same.